roles/v2/mining-proxy/src/lib/upstream_mining.rs

```diff
@@ -110,9 +130,137 @@ impl UpstreamMiningNode {
             downstream_selector,
             group_channels: GroupChannels::new(),
             channel_factory: None,
+            channel_kind,
+            group_id,
+            recv_tp,
+            recv_ph,
+            request_ids,
         }
     }
 
+    pub fn start_receiving_new_template(self_mutex: Arc<Mutex<Self>>) {
+        task::spawn(async move {
+            while self_mutex
+                .safe_lock(|s| s.channel_factory.is_none())
+                .unwrap()
+            {
+                tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+            }
+            let new_template_reciver = self_mutex
+                .safe_lock(|a| a.recv_tp.clone())
+                .unwrap()
+                .unwrap();
+            //let mut res = vec![];
+            loop {
+                let (mut message_new_template, token): (NewTemplate, u64) =
+                    new_template_reciver.recv().await.unwrap();
+                let (channel_id_to_new_job_msg, custom_job) = self_mutex
+                    .safe_lock(|a| {
+                        a.channel_factory
+                            .as_mut()
+                            .unwrap()
+                            .on_new_template(&mut message_new_template)
```

Comment:
This seems like it's doing a lot of work with the lock - this will block every other task (that relies on UpstreamMiningNode) until it's done; is this the intent?

Reply:
You're right, it's not the best thing to do a lot of work here while blocking everything, but the function on_new_template is not blocking even if cpu intensive and right now we do not have performance issues here so I'm going with the simplest implementation. However I'm writing a comment to change this in the future if needed.